src/main/java/org/prebid/server/analytics/reporter/greenbids/GreenbidsAnalyticsReporter.java

```diff
@@ -0,0 +1,341 @@
+    private final HttpClient httpClient;
+    private static final int MAX_16_BIT_INTEGER = 0xFFFF;
+    private static final int RANGE_16_BIT_INTEGER_DIVISION_BASIS = 0x10000;
+
+    public GreenbidsAnalyticsReporter(
+            GreenbidsAnalyticsProperties greenbidsAnalyticsProperties,
+            JacksonMapper jacksonMapper,
+            HttpClient httpClient) {
+        this.greenbidsAnalyticsProperties = Objects.requireNonNull(greenbidsAnalyticsProperties);
+        this.jacksonMapper = Objects.requireNonNull(jacksonMapper);
+        this.httpClient = Objects.requireNonNull(httpClient);
+    }
+
+    @Override
+    public <T> Future<Void> processEvent(T event) {
+        final AuctionContext greenbidsAuctionContext;
+        final BidResponse greenbidsBidResponse;
+
+        if (event instanceof AmpEvent ampEvent) {
+            greenbidsAuctionContext = ampEvent.getAuctionContext();
+            greenbidsBidResponse = ampEvent.getBidResponse();
+        } else if (event instanceof AuctionEvent auctionEvent) {
+            greenbidsAuctionContext = auctionEvent.getAuctionContext();
+            greenbidsBidResponse = auctionEvent.getBidResponse();
+        } else {
+            greenbidsAuctionContext = null;
+            greenbidsBidResponse = null;
+        }
+
+        if (greenbidsBidResponse == null || greenbidsAuctionContext == null) {
+            Future.failedFuture("Bid response or auction context cannot be null");
+        }
+
+        final String greenbidsId = UUID.randomUUID().toString();
+        final String billingId = UUID.randomUUID().toString();
+
+        isSampled(greenbidsAnalyticsProperties.getGreenbidsSampling(), greenbidsId)
+                .compose(isSampled -> {
+                    if (isSampled) {
+                        final Future<CommonMessage> bidMessage = createBidMessage(
+                                greenbidsAuctionContext, greenbidsBidResponse, greenbidsId, billingId);
+
+                        return bidMessage.onSuccess(commonMessage -> {
+                            try {
+                                final String commonMessageJson = jacksonMapper.encodeToString(commonMessage);
+                                HttpUtil.sendJson(commonMessageJson, greenbidsAnalyticsProperties.getAnalyticsServer());
+                            } catch (EncodeException e) {
+                                Future.failedFuture("Failed to encode as JSON: " + e.getMessage());
+                            }
+                        });
+                    }
+                    return null;
+                });
+
+        return Future.succeededFuture();
+    }
+
+    private Future<Boolean> isSampled(
+            double samplingRate,
+            String greenbidsId) {
+        final Promise<Boolean> promise = Promise.promise();
+        try {
+            if (samplingRate < 0 || samplingRate > 1) {
+                Future.failedFuture("Warning: Sampling rate must be between 0 and 1");
```

Comment:
since it's warning it might be converted to the log

fyi This future will be ignored without returning

Reply:
Changed to warning 👍